### PR TITLE
Slicked up database access for Thurloe.

### PR DESCRIPTION
**Pull request by cjllanwarne**

### CODE REVIEW

File: src/main/scala/thurloe/database/ThurloeDatabaseConnector.scala [hunk 1]
```diff
@@ -1,16 +1,88 @@
...
+    val keyValuePairs = TableQuery[DbKeyValuePair]
+    val q = keyValuePairs.filter(x => x.key === key && x.userId === userId)
+    val action = q.delete
+    val affectedRowsCountFuture: Future[Int] = database.run(action)
+
+    // TODO: This is horrible. Don't let this get past a PR if I forget to refactor!
+    val affectedRowCount: Int = Await.result(affectedRowsCountFuture, Duration.fromNanos(500 * 1000 * 1000))
+
+    if (affectedRowCount > 0) Success()
+    else Failure(KeyNotFoundException(userId, key))
```
Comment (geoffjentry) [inline src/main/scala/thurloe/database/ThurloeDatabaseConnector.scala]: General question on these Trys. How are the Failures being used, in particular is the Throwable being used somehow or are you just using the fact that it's a Failure and the error message?

I probably should have pulled this up in non-PR form and traced it but I'm lazy :)

Comment (cjllanwarne) [inline src/main/scala/thurloe/database/ThurloeDatabaseConnector.scala]: They're used in ThurloeService to determine what type of response code to send back.

File: src/main/resources/application.conf [hunk 2]
```diff
@@ -0,0 +1,8 @@
+
+
+h2mem1 = {
+  url = "jdbc:h2:mem:test1"
```
Comment (mcovarr) [inline src/main/resources/application.conf]: Why H2 not not HSQLDB like Cromwell uses?

File: src/main/scala/thurloe/database/ThurloeDatabaseConnector.scala [hunk 3]
```diff
@@ -1,16 +1,103 @@
...
+    setupInMemoryDatabase(database)
+
+  private def lookupWithConstraint(constraint: DbKeyValuePair => Rep[Boolean]): Future[Seq[KeyValuePair]] = {
+    val keyValuePairs = TableQuery[DbKeyValuePair]
+    val q = keyValuePairs.filter(constraint)
+
+    val futureResults = database.run(q.result.transactionally) map {
+      _ map {
+      case (id, userId, key, value) =>
+        println("  " + id + "\t" + userId + "\t" + key + "\t" + value + "\t")
```
Comment (scottfrazer) [inline src/main/scala/thurloe/database/ThurloeDatabaseConnector.scala]: remove println?

File: src/main/scala/thurloe/database/ThurloeDatabaseConnector.scala [hunk 4]
```diff
@@ -1,16 +1,103 @@
...
+  import dataModels.driver.api._
+
+  val database = Database.forConfig("", databaseConfig)
+
+  if (databaseConfig.hasPath("slick.createSchema") && databaseConfig.getBoolean("slick.createSchema"))
+    setupInMemoryDatabase(database)
+
+  private def lookupWithConstraint(constraint: DbKeyValuePair => Rep[Boolean]): Future[Seq[KeyValuePair]] = {
+    val keyValuePairs = TableQuery[DbKeyValuePair]
+    val q = keyValuePairs.filter(constraint)
```
Comment (scottfrazer) [inline src/main/scala/thurloe/database/ThurloeDatabaseConnector.scala]: could probably use a better name than just `q` unless that's some kind of convention I'm not aware of

File: src/main/scala/thurloe/database/ThurloeDatabaseConnector.scala [hunk 5]
```diff
@@ -1,16 +1,103 @@
...
+        Future.failed(InvalidDatabaseStateException(
+          s"Modified $affectedRowsCount rows in database (expected to modify 1)"))
+      }
+    } yield ()
+  }
+
+  def deleteKeyValuePair(userId: String, key: String): Future[Unit] = {
+    val keyValuePairs = TableQuery[DbKeyValuePair]
+    val q = keyValuePairs.filter(x => x.key === key && x.userId === userId)
+    val action = q.delete
```
Comment (scottfrazer) [inline src/main/scala/thurloe/database/ThurloeDatabaseConnector.scala]: maybe `val action = keyValuePairs.filter(x => x.key === key && x.userId === userId).delete`?

Or maybe just rename `val q`?

File: src/main/resources/liquibase/liquibase-deploy.sh [hunk 6]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+liquibase --driver=com.mysql.jdbc.Driver \
+     --classpath=${HOME}/.ivy2/cache/mysql/mysql-connector-java/jars/mysql-connector-java-5.1.35.jar \
+     --changeLogFile=src/main/resources/liquibase/changelog.xml \
+     --url="jdbc:mysql://localhost/thurloe_test" \
+     --username=travis \
+     --password="" \
+     migrate
```
Comment (scottfrazer) [inline src/main/resources/liquibase/liquibase-deploy.sh]: What's this file for?

**Comment by scottfrazer** [comment]
:+1:

File: src/main/scala/thurloe/database/ThurloeDatabaseConnector.scala [hunk 7]
```diff
@@ -1,16 +1,103 @@
...
+        Future.successful(())
+      } else {
+        Future.failed(KeyNotFoundException(userId, key))
+      }
+    } yield ()
+  }
+
+  def setupInMemoryDatabase(database: Database): Unit = {
+    val keyValuePairs = TableQuery[DbKeyValuePair]
+    val setup = DBIO.seq(
```
Comment (geoffjentry) [inline src/main/scala/thurloe/database/ThurloeDatabaseConnector.scala]: Put this on one line

File: src/main/resources/liquibase/liquibase-deploy.sh (hunk 6 above)
Comment (cjllanwarne) [inline src/main/resources/liquibase/liquibase-deploy.sh]: This runs liquibase locally to create the mysql backing database

Comment (mcovarr) [inline src/main/resources/liquibase/liquibase-deploy.sh]: Could that be added to the script as a comment?

Comment (cjllanwarne) [inline src/main/resources/liquibase/liquibase-deploy.sh]: @kshakir do you know whether the liquibase lines in the application.conf make this redundant? Also if we add liquibase to the in-memory DB can we remove the "setupInMemoryDatabase" stuff?

Comment (kshakir) [inline src/main/resources/liquibase/liquibase-deploy.sh]: Whoops. Those liquibase lines are bad copypasta from cromwell. 

Liquibase ci testing isn't in thurloe, yet.

**Pull request merged**